software/scsettings.txt
#######################
# SC1000 config file #
######################
# Any lines beginning with # are ignored


###################
# General Options #
###################
#
# Fader thresholds for hysteresis
# You might need to adjust these for different faders
# fader_open_point should always be
# more than fader_close_point

# Value required to open the fader when it's closed
fader_open_point=5

# Value required to close the fader when it's opened
fader_close_point=3


# 0 to disable platter, default 1 (enabled)
platter_enabled=1

# Ratio of platter movement to sample movement
# 4096 = 1 second for every platter rotation
# 3072 = equivalent to 45rpm record (default)
# 2275 = equivalent to 33rpm
platter_speed=2275

# How long the stop button takes to stop the track
# Higher is longer, default = 3000
brake_speed=3000

# Pitch range % of pitch bend MIDI command, default 50%
pitch_range=50

# How long to delay before looking for MIDI devices
# Useful if your MIDI device takes a long time to initialize
# Or to give you time to switch your phone to MIDI mode
midi_init_delay=5

# How long to delay before looking for audio devices at startup
# Useful if your audio device takes a long time to initialize
audio_init_delay=5

# Set this to 1 or 2 to make the fader cut the beats
# 1 will cut on the left, 2 will cut on the right
cut_beats=0

####################
#   GPIO Mapping   #
####################
#
# This section is used to wire up buttons.
# These can be either directly wired to the Olimex module,
# or on an external MCP23017 GPIO expander
# Format is : 
# gpio=port,pin,pull,edge,action

# port is 0 for MCP23017, or 123456 for ports BCDEFG of A13 on Olimex
# pin number- for MCP23017, A0-A7 is 0-7 and B0-B7 is 8-15
# pull is 1 for pullup, 2 for pulldown
# edge is 1 for pressing edge, 2 for holding action, 3 for pressed while shifted, 4 for held while shifted, 0 for releasing edge
# action is one of :
# 
# (replace x with 0 for beats or 1 for samples) :
# CHx_CUE - Set and trigger cue points
# CHx_SHIFTON - Activate shift mode (ignores x)
# CHx_SHIFTOFF - Deactivate shift mode (also ignores x)
# CHx_STARTSTOP - Start or Stop the track
# CHx_NOTEy - Change the pitch musically, replace y with the note number (60 being 0%)
# CHx_GND - make the pin an output and set it low - used to provide additional ground connections (ignores x)
# gpio=port,pin,pull,edge,action

#### SC500 Buttons ####
# Detection pin
gpio=6,11,1,1,CH0_SC500

# Shift
gpio=2,9,1,1,CH0_SHIFTON
gpio=2,9,1,0,CH0_SHIFTOFF

# Beat Transport
gpio=2,14,1,1,CH0_PREVFILE
gpio=2,14,1,2,CH0_PREVFOLDER
gpio=2,14,1,3,CH0_RANDOMFILE

gpio=2,13,1,1,CH0_STARTSTOP

# even though this is on a beat button it still affects sample
gpio=2,13,1,3,CH1_JOGREVERSE

gpio=2,10,1,1,CH0_NEXTFILE
gpio=2,10,1,2,CH0_NEXTFOLDER
gpio=2,10,1,3,CH0_JOGPIT
gpio=2,10,1,0,CH0_JOGPSTOP

# Beat Volume Buttons
gpio=4,5,1,1,CH0_VOLUP
gpio=4,5,1,2,CH0_VOLUHOLD
gpio=4,4,1,1,CH0_VOLDOWN
gpio=4,4,1,2,CH0_VOLDHOLD

# Sample Transport
gpio=2,5,1,1,CH1_PREVFILE
gpio=2,5,1,2,CH1_PREVFOLDER
gpio=2,5,1,3,CH1_JOGPIT
gpio=2,5,1,0,CH1_JOGPSTOP

gpio=2,6,1,1,CH1_STARTSTOP
gpio=2,6,1,3,CH1_RECORD

gpio=2,8,1,1,CH1_NEXTFILE
gpio=2,8,1,2,CH1_NEXTFOLDER
gpio=2,8,1,3,CH1_RANDOMFILE

# Sample volume buttons
gpio=1,10,1,1,CH1_VOLUP
gpio=1,10,1,2,CH1_VOLUHOLD
gpio=1,4,1,1,CH1_VOLDOWN
gpio=1,4,1,2,CH1_VOLDHOLD

# Sample cue buttons
gpio=2,4,1,1,CH1_CUE
gpio=2,4,1,3,CH1_DELETECUE
gpio=2,7,1,1,CH1_CUE
gpio=2,7,1,3,CH1_DELETECUE
gpio=2,11,1,1,CH1_CUE
gpio=2,11,1,3,CH1_DELETECUE
gpio=2,12,1,1,CH1_CUE
gpio=2,12,1,3,CH1_DELETECUE


#### Expansion port J7 ####
gpio=6,10,1,1,CH0_SHIFTON
gpio=6,10,1,0,CH0_SHIFTOFF
gpio=1,15,1,1,CH0_STARTSTOP
gpio=1,16,1,1,CH1_STARTSTOP

gpio=2,3,1,1,CH1_CUE
gpio=2,0,1,1,CH1_CUE
gpio=2,2,1,1,CH1_CUE
gpio=2,1,1,1,CH1_CUE

gpio=2,3,1,3,CH1_DELETECUE
gpio=2,0,1,3,CH1_DELETECUE
gpio=2,2,1,3,CH1_DELETECUE
gpio=2,1,1,3,CH1_DELETECUE


#### MCP23017 Buttons ####
gpio=0,0,0,1,CH0_GND

gpio=0,1,1,1,CH1_CUE
gpio=0,2,1,1,CH1_CUE
gpio=0,3,1,1,CH1_CUE
gpio=0,4,1,1,CH1_CUE
gpio=0,5,1,1,CH1_CUE
gpio=0,6,1,1,CH1_CUE
gpio=0,7,1,1,CH1_CUE

gpio=0,1,1,3,CH1_DELETECUE
gpio=0,2,1,3,CH1_DELETECUE
gpio=0,3,1,3,CH1_DELETECUE
gpio=0,4,1,3,CH1_DELETECUE
gpio=0,5,1,3,CH1_DELETECUE
gpio=0,6,1,3,CH1_DELETECUE
gpio=0,7,1,3,CH1_DELETECUE

gpio=0,8,1,1,CH1_STARTSTOP
gpio=0,9,1,1,CH0_SHIFTON
gpio=0,9,1,0,CH0_SHIFTOFF 
gpio=0,10,1,1,CH1_CUE
gpio=0,10,1,3,CH1_DELETECUE
gpio=0,11,1,1,CH0_STARTSTOP
gpio=0,12,0,1,CH1_GND
gpio=0,13,1,1,CH1_CUE
gpio=0,14,1,1,CH1_CUE
gpio=0,15,1,1,CH1_CUE
gpio=0,13,1,3,CH1_DELETECUE
gpio=0,14,1,3,CH1_DELETECUE
gpio=0,15,1,3,CH1_DELETECUE



################
# MIDI Mapping #
################
#
# Format is :
# Status,channel,data,edge, action
# 
# Status can be : 
# 8 - Note OFF (will also match zero-velocity Note ON)
# 9 - Note ON
# 11 - Control Change
# 14 - Pitch Bend
#
# Channel is the MIDI channel of the incoming MIDI message
#
# Data is the note or CC number - set to "255" to bind to ALL notes of a particular channel
#
# Action can be (replace x with 0 for beats or 1 for samples) :
# CHx_CUE - Set and trigger cue points
# CHx_SHIFTON - Activate shift mode (ignores x)
# CHx_SHIFTOFF - Deactivate shift mode (also ignores x)
# CHx_STARTSTOP - Start or Stop the track
# CHx_PITCH - Change the speed of the track
# CHx_NOTEy - Change the pitch musically, replace y with the note number (60 being 0%)



#### LEMUR PRESET, normally hardcoded ####

# CC 0 of channels 0/1 is volume
midii=11,0,0,1,CH0_VOLUME
midii=11,1,0,1,CH1_VOLUME

# Pitch bend on channels 0 and 1 is pitch

midii=14,0,0,1,CH0_PITCH
midii=14,1,0,1,CH1_PITCH

# notes on channels 0 and 1 are cue points
midii=9,0,255,1,CH0_CUE
midii=9,1,255,1,CH1_CUE


# Also add the delete cue command for shift modifier
midii=9,0,255,3,CH0_DELETECUE
midii=9,1,255,3,CH1_DELETECUE


# Notes on channel 3 are C1-style notes
midii=9,3,255,1,CH1_NOTE


# Channel 4 has various other commands mapped to its notes
# Notes 0-1 = start stop
midii=9,4,0,1,CH0_STARTSTOP
midii=9,4,1,1,CH1_STARTSTOP

# Notes 2-3 = Next file
midii=9,4,2,1,CH0_NEXTFILE
midii=9,4,3,1,CH1_NEXTFILE

# Notes 4-5 = Next folder
midii=9,4,4,1,CH0_NEXTFOLDER
midii=9,4,5,1,CH1_NEXTFOLDER

# Notes 6-7 = Prev file
midii=9,4,6,1,CH0_PREVFILE
midii=9,4,7,1,CH1_PREVFILE

# Notes 8-9 = Prev folder
midii=9,4,8,1,CH0_PREVFOLDER
midii=9,4,9,1,CH1_PREVFOLDER

# Notes 10-11 = Random file
midii=9,4,10,1,CH0_RANDOMFILE
midii=9,4,11,1,CH1_RANDOMFILE

# Note 126 = record
midii=9,4,126,1,CH0_RECORD

# Note 127 = shift
midii=9,4,127,1,CH1_SHIFTON
midii=8,4,127,3,CH1_SHIFTOFF

# Status,channel,data,edge,action

software/src/app/settings.c
//
// Created by lodsb on 22-Mar-25.
//

#include <unistd.h>         //Needed for I2C port
#include <fcntl.h>         //Needed for I2C port
#include <sys/ioctl.h>      //Needed for I2C port

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "../global/global.h"
#include "../input/sc_midimap.h"

#include "settings.h"

unsigned int count_chars( char* string, char c )
{
   unsigned int count = 0;

   //printf("Checking for commas in %s\n", string);

   do
   {
      if ( (*string) == c )
      {
         count++;
      }
   } while ( (*(string++)) );
   return count;
}

void settings_load_user_configuration( struct sc_settings* settings, struct mapping* maps )
{
   FILE* fp;
   char* line = NULL;
   size_t len = 0;
   ssize_t read;
   char* param, * actions;
   char* value;
   unsigned char channel = 0, notenum = 0, controlType = 0, pin = 0, pullup = 0, port = 0;
   char edge;
   char delim[] = "=";
   char delimc[] = ",";
   unsigned char midicommand[3];
   char* linetok, * valuetok;
   // set defaults
   settings->period_size = 256;
   settings->buffer_period_factor = 4;
   settings->fader_close_point = 2;
   settings->fader_open_point = 10;
   settings->platter_enabled = 1;
   settings->platter_speed = 2275;
   settings->sample_rate = 48000;
   settings->update_rate = 2000;
   settings->debounce_time = 5;
   settings->hold_time = 100;
   settings->slippiness = 200;
   settings->brake_speed = 3000;
   settings->pitch_range = 50;
   settings->midi_init_delay = 5;
   settings->audio_init_delay = 2;
   settings->volume_amount = 0.03;
   settings->volume_amount_held = 0.001;
   settings->initial_volume = 0.125;
   settings->midi_remapped = 0;
   settings->io_remapped = 0;
   settings->jog_reverse = 0;
   settings->cut_beats = 0;
   settings->importer = DEFAULT_IMPORTER;

   // later we'll check for sc500 pin and use it to set following settings
   settings->disable_volume_adc = 0;
   settings->disable_pic_buttons = 0;

   // Load any settings from config file
   fp = fopen("/media/sda/scsettings.txt", "r");
   if ( fp == NULL )
   {
      // load internal copy instead
      fp = fopen("/var/scsettings.txt", "r");
   }


   if ( fp != NULL )
   {
      while ( (read = getline(&line, &len, fp)) != -1 )
      {
         if ( strlen(line) < 2 || line[ 0 ] == '#' )
         { // Comment or blank line
         }
         else
         {
            param = strtok_r(line, delim, &linetok);
            value = strtok_r(NULL, delim, &linetok);

            if ( strcmp(param, "period_size") == 0 )
            {
               settings->period_size = atoi(value);
            }
            else if ( strcmp(param, "buffer_period_factor") == 0 )
            {
               settings->buffer_period_factor = atoi(value);
            }
            else if ( strcmp(param, "fader_close_point") == 0 )
            {
               settings->fader_close_point = atoi(value);
            }
            else if ( strcmp(param, "fader_open_point") == 0 )
            {
               settings->fader_open_point = atoi(value);
            }
            else if ( strcmp(param, "platter_enabled") == 0 )
            {
               settings->platter_enabled = atoi(value);
            }
            else if ( strcmp(param, "disable_volume_adc") == 0 )
            {
               settings->disable_volume_adc = atoi(value);
            }
            else if ( strcmp(param, "platter_speed") == 0 )
            {
               settings->platter_speed = atoi(value);
            }
            else if ( strcmp(param, "sample_rate") == 0 )
            {
               settings->sample_rate = atoi(value);
            }
            else if ( strcmp(param, "update_rate") == 0 )
            {
               settings->update_rate = atoi(value);
            }
            else if ( strcmp(param, "debounce_time") == 0 )
            {
               settings->debounce_time = atoi(value);
            }
            else if ( strcmp(param, "hold_time") == 0 )
            {
               settings->hold_time = atoi(value);
            }
            else if ( strcmp(param, "slippiness") == 0 )
            {
               settings->slippiness = atoi(value);
            }
            else if ( strcmp(param, "brake_speed") == 0 )
            {
               settings->brake_speed = atoi(value);
            }
            else if ( strcmp(param, "pitch_range") == 0 )
            {
               settings->pitch_range = atoi(value);
            }
            else if ( strcmp(param, "jog_reverse") == 0 )
            {
               settings->jog_reverse = atoi(value);
            }
            else if ( strcmp(param, "cut_beats") == 0 )
            {
               settings->cut_beats = atoi(value);
            }
            else if ( strstr(param, "midii") != NULL )
            {
               settings->midi_remapped = 1;
               controlType = atoi(strtok_r(value, delimc, &valuetok));
               channel = atoi(strtok_r(NULL, delimc, &valuetok));
               notenum = atoi(strtok_r(NULL, delimc, &valuetok));
               edge = atoi(strtok_r(NULL, delimc, &valuetok));
               actions = strtok_r(NULL, delimc, &valuetok);

               //255 means bind to all notes
               if ( notenum == 255 )
               {

                  // Build MIDI command
                  midicommand[ 0 ] = (controlType << 4) | channel;
                  midicommand[ 1 ] = notenum;
                  midicommand[ 2 ] = 0;

                  char tempact[20];
                  for ( midicommand[ 1 ] = 0; midicommand[ 1 ] < 128; midicommand[ 1 ]++ )
                  {
                     sprintf(tempact, "%s%u", actions, midicommand[ 1 ]);
                     add_config_mapping(
                             &maps,
                             MAP_MIDI,
                             midicommand,
                             0,
                             0,
                             0,
                             edge,
                             tempact);
                  }
               }
                  // otherwise just bind to one note
               else
               {
                  // Build MIDI command
                  midicommand[ 0 ] = (controlType << 4) | channel;
                  midicommand[ 1 ] = notenum;
                  midicommand[ 2 ] = 0;
                  add_config_mapping(
                          &maps,
                          MAP_MIDI,
                          midicommand,
                          0,
                          0,
                          0,
                          edge,
                          actions);
               }
            }
            else if ( strstr(param, "io") != NULL )
            {
               settings->io_remapped = 1;
               unsigned int commaCount = count_chars(value, ',');
               //printf("Found io %s - comacount %d\n", value, commaCount);
               port = 0;
               if ( commaCount == 4 )
               {
                  port = atoi(strtok_r(value, delimc, &valuetok));
                  pin = atoi(strtok_r(NULL, delimc, &valuetok));
               }
               else
               {
                  pin = atoi(strtok_r(value, delimc, &valuetok));
               }
               pullup = atoi(strtok_r(NULL, delimc, &valuetok));
               edge = atoi(strtok_r(NULL, delimc, &valuetok));
               actions = strtok_r(NULL, delimc, &valuetok);
               add_config_mapping(
                       &maps,
                       MAP_IO,
                       NULL,
                       port,
                       pin,
                       pullup,
                       edge,
                       actions);
            }
            else if ( strcmp(param, "midi_init_delay") == 0 )
            {// Literally just a sleep to allow USB devices longer to initialize
               settings->midi_init_delay = atoi(value);
            }
            else if ( strcmp(param, "audio_init_delay") == 0 )
            {
               fprintf(stderr, "audio init settings");
               // Literally just a sleep to allow USB devices longer to initialize
               settings->audio_init_delay = atoi(value);
            }
            else
            {
               printf("Unrecognised configuration line - Param : %s , value : %s\n", param, value);
            }
         }
      }
   }


   printf("ps %d, bpf %d, fcp %d, fop %d, pe %d, ps %d, sr %d, ur %d\n",
          settings->period_size,
          settings->buffer_period_factor,
          settings->fader_close_point,
          settings->fader_open_point,
          settings->platter_enabled,
          settings->platter_speed,
          settings->sample_rate,
          settings->update_rate);

   if ( fp )
   {
      fclose(fp);
   }
   if ( line )
   {
      free(line);
   }
}